Camera plugin crashes Android 9 device

<!--
Please have a look at the issue templates you get when you click "New issue" in the GitHub UI.
We very much prefer issues created by using one of these templates.
-->

### Issue Type
<!-- Please check the boxes by putting an x in the [ ] like so: [x] -->

- [x] Bug Report
- [ ] Feature Request
- [ ] Support Question

## Description
The camera plugin randomly crashes my Ionic v1 app when taking a picture with the camera. Sometimes after taking 50 pictures, sometimes after taking 5 pictures. 

## Information
See next comments for the stacktrace and full logging of the app crash.
Its pretty difficult to provide a reproducable case since the crashes happen randomly.
The best way to reproduce is to simply create an empty Ionic v1 project, add the plugin and start taking pictures on an **Android 9** device.

### Command or Code
See comments below for logging

Code:
            navigator.camera.getPicture(deferred.resolve, function(error) {
                $log.error("Unable to obtain picture: ", error);
                deferred.reject(error);
            }, {
                allowEdit: false,
                correctOrientation: true,
                destinationType: Camera.DestinationType.FILE_URI,
                direction: Camera.Direction.BACK,
                encodingType: Camera.EncodingType.JPEG,
                mediaType: Camera.MediaType.PICTURE,
                quality: 100,
                sourceType: Camera.PictureSourceType.CAMERA,
                targetWidth: 1000,
                targetHeight: 625
            });
<!-- What command or code is needed to reproduce the problem? -->

### Environment, Platform, Device

- Environment: Ionic v1
- Ionic CLI: 2.1.17
- Cordova CLI: 7.1.0
- Platform: Android
- Camera plugin version: 4.1.0
- Device: Samsung Galaxy tab S5e(Android 9)
- Cordova-android version: 8.0.0
- Npm CLI: 5.6.0
- Node version: 8.11.3


## Checklist
<!-- Please check the boxes by putting an `x` in the `[ ]` like so: `[x]` -->

- [x] I searched for already existing GitHub issues about this
- [ ] I updated all Cordova tooling to their most recent version
- [x] I included all the necessary information above

**Stacktrace of error occured during the crash:**

01-28 11:47:39.946 10440-10467/? E/PdkUtil: Fail to acquire available samsung key list.
    java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Method.invoke(Native Method)
        at com.samsung.android.camera.core2.local.internal.PdkUtil.getAvailableSamsungKeyList(PdkUtil.java:313)
        at com.samsung.android.camera.core2.local.vendorkey.SemCameraCharacteristics.getAvailableSessionKeys(SemCameraCharacteristics.java:1004)
        at com.samsung.android.camera.core2.CamCapability.getAvailableSamsungSessionKeys(CamCapability.java:528)
        at com.samsung.android.camera.core2.CamCapability.getAvailableSamsungSessionKeyNames(CamCapability.java:553)
        at com.samsung.android.camera.core2.maker.PhotoMakerBase.connectCamDevice(PhotoMakerBase.java:992)
        at com.samsung.android.camera.core2.maker.ProcessingPhotoMakerBase.connectCamDevice(ProcessingPhotoMakerBase.java:66)
        at com.sec.android.app.camera.engine.request.ConnectMakerRequest.execute(ConnectMakerRequest.java:102)
        at com.sec.android.app.camera.engine.request.Request.run(Request.java:88)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalArgumentException: Could not find tag for key 'samsung.android.request.availableSessionKeys')
        at android.hardware.camera2.impl.CameraMetadataNative.nativeGetTagFromKeyLocal(Native Method)
        at android.hardware.camera2.impl.CameraMetadataNative.getBase(CameraMetadataNative.java:514)
        at android.hardware.camera2.impl.CameraMetadataNative.get(CameraMetadataNative.java:404)
        at android.hardware.camera2.impl.CameraMetadataNative.get(CameraMetadataNative.java:373)
        at android.hardware.camera2.CameraCharacteristics.get(CameraCharacteristics.java:223)
        at android.hardware.camera2.CameraCharacteristics.getAvailableSamsungKeyList(CameraCharacteristics.java:287)
        at com.samsung.android.sdk.camera.impl.internal.CustomInterfaceHelper.getAvailableSamsungKeyList(CustomInterfaceHelper.java:41)
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.samsung.android.camera.core2.local.internal.PdkUtil.getAvailableSamsungKeyList(PdkUtil.java:313) 
        at com.samsung.android.camera.core2.local.vendorkey.SemCameraCharacteristics.getAvailableSessionKeys(SemCameraCharacteristics.java:1004) 
        at com.samsung.android.camera.core2.CamCapability.getAvailableSamsungSessionKeys(CamCapability.java:528) 
        at com.samsung.android.camera.core2.CamCapability.getAvailableSamsungSessionKeyNames(CamCapability.java:553) 
        at com.samsung.android.camera.core2.maker.PhotoMakerBase.connectCamDevice(PhotoMakerBase.java:992) 
        at com.samsung.android.camera.core2.maker.ProcessingPhotoMakerBase.connectCamDevice(ProcessingPhotoMakerBase.java:66) 
        at com.sec.android.app.camera.engine.request.ConnectMakerRequest.execute(ConnectMakerRequest.java:102) 
        at com.sec.android.app.camera.engine.request.Request.run(Request.java:88) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:764) 
01-28 11:47:39.983 681-10479/? E/CHI: [SS_ERR ]: [CHI         ]: chxusecaseutils.cpp: Destroy: 1706: ImageBuffer 0xe4287da0 is destroyed with reference count = 1

I see the same problem on our app.
On android 8 i see no problems. On android 9 it crashes randomly.
To reproduce i need to shoot a lot of photos, but i have reports from users that has the problem more often.
The problem is with cordova-camera 4.0.3 and also with 4.1.0. I dint tried lower versions.

In other places i read thet my stacktrace could be about permission on disk, but its seem all good. On normal use i save the images correctly, and some rare times i cant. So the permissions are right.

My code to call the camera :
`navigator.camera.getPicture(onSuccess, onFail, {
                quality: 50,
                allowEdit: true,
                destinationType: Camera.DestinationType.FILE_URI,
                correctOrientation: true
            });`

Here the trace : 
`03-13 10:45:10.585 16037 16037 D AndroidRuntime: Shutting down VM
03-13 10:45:10.586 16037 16037 E AndroidRuntime: FATAL EXCEPTION: main
03-13 10:45:10.586 16037 16037 E AndroidRuntime: Process: it.tideup.riquesto, PID: 16037
03-13 10:45:10.586 16037 16037 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {it.tideup.riquesto/it.tideup.riquesto.MainActivity}: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=34, result=-1, data=null} to activity {it.tideup.riquesto/it.tideup.riquesto.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.res.XmlResourceParser android.content.pm.ProviderInfo.loadXmlMetaData(android.content.pm.PackageManager, java.lang.String)' on a null object reference
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3830)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3862)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:51)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:145)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1816)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:106)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:193)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6854)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:860)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: Caused by: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=34, result=-1, data=null} to activity {it.tideup.riquesto/it.tideup.riquesto.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.res.XmlResourceParser android.content.pm.ProviderInfo.loadXmlMetaData(android.content.pm.PackageManager, java.lang.String)' on a null object reference
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4427)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3819)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	... 11 more
03-13 10:45:10.586 16037 16037 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.res.XmlResourceParser android.content.pm.ProviderInfo.loadXmlMetaData(android.content.pm.PackageManager, java.lang.String)' on a null object reference
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.support.v4.content.FileProvider.parsePathStrategy(FileProvider.java:604)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.support.v4.content.FileProvider.getPathStrategy(FileProvider.java:578)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.support.v4.content.FileProvider.getUriForFile(FileProvider.java:416)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at org.apache.cordova.camera.CameraLauncher.onActivityResult(CameraLauncher.java:803)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at org.apache.cordova.CordovaInterfaceImpl.onActivityResult(CordovaInterfaceImpl.java:152)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at org.apache.cordova.CordovaActivity.onActivityResult(CordovaActivity.java:358)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.Activity.dispatchActivityResult(Activity.java:7469)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	at android.app.ActivityThread.deliverResults(ActivityThread.java:4420)
03-13 10:45:10.586 16037 16037 E AndroidRuntime: 	... 12 more`

--- a/src/android/FileHelper.java
+++ b/src/android/FileHelper.java
@@ -355,5 +355,4 @@ public class FileHelper {
         return file.exists() ? file.toString(): null;
     }
 
-
 }

--- a/src/android/FileHelper.java
+++ b/src/android/FileHelper.java
@@ -19,7 +19,6 @@ package org.apache.cordova.camera;
 import android.annotation.SuppressLint;
 import android.content.ContentUris;
 import android.content.Context;
-import android.content.CursorLoader;
 import android.database.Cursor;
 import android.net.Uri;
 import android.os.Build;
@@ -29,8 +28,8 @@ import android.provider.MediaStore;
 import android.webkit.MimeTypeMap;
 
 import org.apache.cordova.CordovaInterface;
-import org.apache.cordova.LOG;
 
+import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
@@ -44,7 +43,7 @@ public class FileHelper {
      * Returns the real path of the given URI string.
      * If the given URI string represents a content:// URI, the real path is retrieved from the media store.
      *
-     * @param uriString the URI string of the audio/image/video
+     * @param uri the URI of the audio/image/video
      * @param cordova the current application context
      * @return the full path to the file
      */
@@ -66,7 +65,7 @@ public class FileHelper {
      * Returns the real path of the given URI.
      * If the given URI is a content:// URI, the real path is retrieved from the media store.
      *
-     * @param uri the URI of the audio/image/video
+     * @param uriString the URI string from which to obtain the input stream
      * @param cordova the current application context
      * @return the full path to the file
      */
@@ -143,6 +142,9 @@ public class FileHelper {
             if (isGooglePhotosUri(uri))
                 return uri.getLastPathSegment();
 
+            if (isFileProviderUri(context, uri))
+                return getFileProviderPath(context, uri);
+
             return getDataColumn(context, uri, null, null);
         }
         // File
@@ -188,6 +190,7 @@ public class FileHelper {
             if (question > -1) {
                 uriString = uriString.substring(0, question);
             }
+
             if (uriString.startsWith("file:///android_asset/")) {
                 Uri uri = Uri.parse(uriString);
                 String relativePath = uri.getPath().substring(15);
@@ -217,6 +220,7 @@ public class FileHelper {
      * @return a path without the "file://" prefix
      */
     public static String stripFileProtocol(String uriString) {
+
         if (uriString.startsWith("file://")) {
             uriString = uriString.substring(7);
         }
@@ -327,4 +331,29 @@ public class FileHelper {
     public static boolean isGooglePhotosUri(Uri uri) {
         return "com.google.android.apps.photos.content".equals(uri.getAuthority());
     }
+
+    /**
+     * @param context The Application context
+     * @param uri The Uri is checked by functions
+     * @return Whether the Uri authority is FileProvider
+     */
+    public static boolean isFileProviderUri(final Context context, final Uri uri) {
+        final String packageName = context.getPackageName();
+        final String authority = new StringBuilder(packageName).append(".provider").toString();
+        return authority.equals(uri.getAuthority());
+    }
+
+    /**
+     * @param context The Application context
+     * @param uri The Uri is checked by functions
+     * @return File path or null if file is missing
+     */
+    public static String getFileProviderPath(final Context context, final Uri uri)
+    {
+        final File appDir = context.getExternalFilesDir(Environment.DIRECTORY_PICTURES);
+        final File file = new File(appDir, uri.getLastPathSegment());
+        return file.exists() ? file.toString(): null;
+    }
+
+
 }

